src/fabric.c

```diff
@@ -481,6 +481,27 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node, const cha
 
 	*info = tail = NULL;
 	for (prov = prov_head; prov; prov = prov->next) {
+		/* If FI_PROV_ATTR_ONLY is specified, make a dummy
+		 * info object for each provider, and copy in the
+		 * provider name and version */
+		if (flags == FI_PROV_ATTR_ONLY) {
+			cur = fi_allocinfo();
+			cur->fabric_attr->prov_name =
+				strdup(prov->provider->name);
+			cur->fabric_attr->prov_version =
+				prov->provider->version;
+
+			if (!*info) {
+				*info = tail = cur;
+			} else {
+				tail->next = cur;
+			}
+			tail = cur;
+
+			ret = 0;
+			continue;
+		}
+
```

Comment:
Can you move this into its own function?  I.e. check for the flag outside of the loop and call fi_getprovinfo() or something like that?